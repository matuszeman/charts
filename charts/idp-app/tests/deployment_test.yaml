suite: deployment
templates:
  - deployment.yaml
values:
  - values/minimum.yaml
tests:
  - it: should match snapshot with values-ci.yaml
    values:
      - ../values-ci.yaml
    set:
      containers:
        app:
          imageRepository: private
    asserts:
      - matchSnapshot: {}
  - it: when nodePool is set
    values:
      - values/deployment.yaml
    set:
      nodePool: test1
    asserts:
      - isSubset:
          path: spec.template.spec.nodeSelector
          content:
            environment: test
      - contains:
          path: spec.template.spec.tolerations
          content:
            effect: NoSchedule
            key: environment
            operator: Equal
            value: test
  - it: when terminationGracePeriodSeconds is set
    values:
      - values/deployment.yaml
    set:
      terminationGracePeriodSeconds: 120
    asserts:
      - equal:
          path: spec.template.spec.terminationGracePeriodSeconds
          value: 120
  - it: can set nodeSelector manually
    values:
      - values/deployment.yaml
    set:
      nodeSelector:
        karpenter.k8s.aws/instance-category: c
    asserts:
      - isSubset:
          path: spec.template.spec.nodeSelector
          content:
            karpenter.k8s.aws/instance-category: c
  - it: private image repo is default
    values:
      - values/deployment.yaml
    set:
      containers:
        app:
          image: test-image
    asserts:
      - equal:
          path: spec.template.spec.containers[0].image
          value: 123456.dkr.ecr.eu-central-1.amazonaws.com/test-image:0.1.0
  - it: public image repo
    values:
      - values/deployment.yaml
    set:
      containers:
        app:
          image: public-image
          imageRepository: public
    asserts:
      - equal:
          path: spec.template.spec.containers[0].image
          value: public.ecr.aws/123456/public-image:0.1.0
  - it: deployment strategy Recreate
    values:
      - values/deployment.yaml
    set:
      deployment:
        strategy:
          type: Recreate
    asserts:
      - equal:
          path: spec.strategy
          value:
            type: Recreate
  - it: when no deployment strategy is defined
    values:
      - values/deployment.yaml
    set:
      deployment:
        strategy: null
    asserts:
      - notExists:
          path: spec.strategy
  - it: templated appVersion
    values:
      - values/deployment.yaml
    set:
      global:
        appVersion: "VERSION"
      appVersion: "{{ .Values.global.appVersion }}"
    asserts:
      - equal:
          path: metadata.labels["app.kubernetes.io/version"]
          value: VERSION
  - it: templated container imageTag
    values:
      - values/deployment.yaml
    set:
      global:
        appVersion: "VERSION"
      containers:
        app:
          imageTag: "{{ .Values.global.appVersion }}"
    asserts:
      - equal:
          path: spec.template.spec.containers[0].image
          value: 123456.dkr.ecr.eu-central-1.amazonaws.com/test-server:VERSION
  - it: dockerhub image repo
    values:
      - values/deployment.yaml
    set:
      containers:
        app:
          image: username/image
          imageRepository: docker.io
    asserts:
      - equal:
          path: spec.template.spec.containers[0].image
          value: docker.io/username/image:0.1.0
  - it: command and args
    values:
      - values/deployment.yaml
    set:
      containers:
        app:
          command:
            - customcmd
            - 567
          args:
            - 123
            - string
    asserts:
      - equal:
          path: spec.template.spec.containers[0].command
          value:
            - customcmd
            - "567"  # commands are quoted
      - equal:
          path: spec.template.spec.containers[0].args
          value:
            - "123" # args are quoted
            - string
  - it: when another non app container is provided, then it does not have to provide containerPort
    values:
      - values/deployment.yaml
    set:
      containers:
        logging:
          image: logging
          imageTag: 1.0.0
          resources:
            limits:
              cpu: 200m
              memory: 256Mi
            requests:
              cpu: 100m
              memory: 128Mi
    asserts:
      - equal:
          path: spec.template.spec.containers[1].name
          value: logging

  - it: when resources.limits are defined
    values:
      - values/deployment.yaml
    set:
      containers:
        app:
          resources:
            limits:
              cpu: 200m
              memory: 200Mi
              ephemeralStorage: 100Mi
    asserts:
      - equal:
          path: spec.template.spec.containers[0].resources.limits
          value:
            cpu: 200m
            memory: 200Mi
            ephemeral-storage: 100Mi

  - it: when resources.requests.ephemeralStorage is defined
    values:
      - values/deployment.yaml
    set:
      containers:
        app:
          resources:
            requests:
              ephemeralStorage: 50Mi
    asserts:
      - isSubset:
          path: spec.template.spec.containers[0].resources.requests
          content:
            ephemeral-storage: 50Mi

  - it: when multi deployment
    values:
      - values/deployment.yaml
      - values/multi.yaml

    asserts:
      - matchSnapshot: {}
      - equal:
          path: metadata.name
          value: RELEASE-NAME-one
        documentIndex: 0
      - equal:
          path: metadata.name
          value: RELEASE-NAME-two
        documentIndex: 1

  - it: when multi deployment with containers override
    values:
      - values/deployment.yaml
    set:
      containers:
        app:
          command: ["command", "original"]
      deployment:
        multi:
          one: {}
          two:
            containers:
              app:
                command: ["command", "override"]

    asserts:
      - equal:
          path: spec.template.spec.containers[0].command
          value: ["command", "original"]
        documentIndex: 0
      - equal:
          path: spec.template.spec.containers[0].command
          value: ["command", "override"]
        documentIndex: 1

  - it: when containerPortProtocol is defined
    values:
      - values/deployment.yaml
    set:
      containers:
        app:
          containerPortProtocol: UDP
    asserts:
      - equal:
          path: spec.template.spec.containers[0].ports[0].protocol
          value: UDP
  - it: when deployment.kind is StatefulSet
    values:
      - values/deployment.yaml
      - values/statefulset.yaml
    asserts:
      - equal:
          path: kind
          value: StatefulSet
  - it: default restartPolicy is undefined
    values:
      - values/deployment.yaml
    asserts:
      - notExists:
          path: spec.template.spec.restartPolicy
  - it: when annotations are set
    values:
      - values/deployment.yaml
    set:
      deployment:
        annotations:
          some: annot
    asserts:
      - isSubset:
          path: metadata.annotations
          content:
            some: annot
  - it: when enableServiceLinks are set to true, then do not render it as it's kube default
    values:
      - values/deployment.yaml
    set:
      enableServiceLinks: true
    asserts:
      - notExists:
          path: spec.template.spec.enableServiceLinks