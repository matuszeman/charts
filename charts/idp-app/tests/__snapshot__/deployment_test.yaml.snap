should match snapshot with values-ci.yaml:
  1: |
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      labels:
        app.kubernetes.io/instance: RELEASE-NAME-resource-suffix
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: app-name
        app.kubernetes.io/version: 0.1.0
      name: RELEASE-NAME-resource-suffix
    spec:
      revisionHistoryLimit: 1
      selector:
        matchLabels:
          app.kubernetes.io/instance: RELEASE-NAME-resource-suffix
          app.kubernetes.io/name: app-name
      template:
        metadata:
          labels:
            app.kubernetes.io/instance: RELEASE-NAME-resource-suffix
            app.kubernetes.io/name: app-name
        spec:
          affinity:
            podAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                - labelSelector:
                    matchExpressions:
                      - key: security
                        operator: In
                        values:
                          - S1
                  topologyKey: topology.kubernetes.io/zone
            podAntiAffinity:
              preferredDuringSchedulingIgnoredDuringExecution:
                - podAffinityTerm:
                    labelSelector:
                      matchExpressions:
                        - key: security
                          operator: In
                          values:
                            - S2
                    topologyKey: topology.kubernetes.io/zone
                  weight: 100
          containers:
            - image: 123456.dkr.ecr.eu-central-1.amazonaws.com/idp-example-web-app-apache:0.1.0
              imagePullPolicy: IfNotPresent
              livenessProbe:
                httpGet:
                  path: /
                  port: apache
              name: apache
              ports:
                - containerPort: 80
                  name: apache
                  protocol: TCP
              readinessProbe:
                httpGet:
                  path: /
                  port: apache
              resources:
                limits:
                  cpu: 200m
                  memory: 256Mi
                requests:
                  cpu: 100m
                  memory: 128Mi
            - args:
                - --quoted=arg1
                - --quoted arg2
              command:
                - customcmdquoted
              env:
                - name: configFromAwsSecret
                  valueFrom:
                    secretKeyRef:
                      key: somekey
                      name: RELEASE-NAME-resource-suffix-config-app-secrets
                - name: configFromConfigContent
                  valueFrom:
                    configMapKeyRef:
                      key: plaintextkey
                      name: RELEASE-NAME-resource-suffix-config-app
                - name: configFromConfigMap
                  valueFrom:
                    configMapKeyRef:
                      key: somekey
                      name: RELEASE-NAME-not-managed-cm
                - name: configFromConfigMapHashed
                  valueFrom:
                    configMapKeyRef:
                      key: somekey
                      name: RELEASE-NAME-not-managed-cm-hashed-4d4c3f1
                - name: configFromSecret
                  valueFrom:
                    secretKeyRef:
                      key: somekey
                      name: RELEASE-NAME-not-managed-secret
                - name: configFromSecretHashed
                  valueFrom:
                    secretKeyRef:
                      key: somekey
                      name: RELEASE-NAME-not-managed-secret-hashed-4aef4fd
                - name: configKeyTemplatedValue
                  valueFrom:
                    configMapKeyRef:
                      key: cm-key
                      name: app-config-RELEASE-NAME
                - name: configKeyValue
                  valueFrom:
                    configMapKeyRef:
                      key: cm-key
                      name: app-config
                - name: floatValue
                  value: "456.789"
                - name: intValue
                  value: "123"
                - name: secretKeyTemplatedValue
                  valueFrom:
                    secretKeyRef:
                      key: secret-key
                      name: app-secrets-RELEASE-NAME
                - name: secretKeyValue
                  valueFrom:
                    secretKeyRef:
                      key: some-key
                      name: app-secrets
                - name: stringValue
                  value: string
                - name: templatedValue
                  value: value-RELEASE-NAME
              image: 123456.dkr.ecr.eu-central-1.amazonaws.com/test-server:0.1.0
              imagePullPolicy: IfNotPresent
              lifecycle:
                preStop:
                  exec:
                    command:
                      - /bin/sh
                      - -c
                      - echo 'preStop hook'
              livenessProbe:
                httpGet:
                  path: /
                  port: app
              name: app
              ports:
                - containerPort: 9000
                  name: app
                  protocol: TCP
                - containerPort: 9000
                  name: anotherport
              readinessProbe:
                httpGet:
                  path: /
                  port: app
              resources:
                limits:
                  cpu: 200m
                  memory: 256Mi
                requests:
                  cpu: 100m
                  memory: 128Mi
              startupProbe:
                failureThreshold: 30
                httpGet:
                  path: /
                  port: app
                periodSeconds: 10
              volumeMounts:
                - mountPath: /app-config
                  name: config-app
                  readOnly: true
                - mountPath: /app-secrets
                  name: config-app-secrets
                  readOnly: true
                - mountPath: /app-secrets-manual
                  name: config-app-secrets-manual
                  readOnly: true
                - mountPath: /hashed-aws-secret
                  name: config-hashed-aws-secret
                  readOnly: true
                - mountPath: /hashed-cm
                  name: config-hashed-cm
                  readOnly: true
                - mountPath: /not-managed-secret
                  name: config-not-managed-secret
                  readOnly: true
                - mountPath: /scripts
                  name: config-scripts
                  readOnly: true
                - mountPath: /tmp/example.txt
                  name: config-not-managed-cm
                  readOnly: true
                  subPath: example.txt
                - mountPath: /etc/php.ini
                  name: config-php
                  readOnly: true
                  subPath: php.ini
                - mountPath: /app/public/var
                  name: ephemeral-app-cache
                  readOnly: false
                - mountPath: /app/public/readonly-by-default.yaml
                  name: some-files
                  readOnly: true
                  subPath: readonly-by-default.yaml
                - mountPath: /app/public/readonly.yaml
                  name: some-files
                  readOnly: true
                  subPath: readonly.yaml
                - mountPath: /app/public/writable.yaml
                  name: some-files
                  readOnly: false
                  subPath: writable.yaml
            - image: 123456.dkr.ecr.eu-central-1.amazonaws.com/logging:1.0.0
              imagePullPolicy: IfNotPresent
              name: logging
              resources:
                limits:
                  cpu: 200m
                  memory: 256Mi
                requests:
                  cpu: 100m
                  memory: 128Mi
          enableServiceLinks: false
          nodeSelector:
            environment: test
          securityContext:
            sysctls:
              - name: net.ipv4.ip_unprivileged_port_start
                value: "0"
          serviceAccountName: RELEASE-NAME-resource-suffix
          tolerations:
            - effect: NoSchedule
              key: environment
              operator: Equal
              value: test
          topologySpreadConstraints:
            - labelSelector:
                matchLabels:
                  app.kubernetes.io/instance: RELEASE-NAME-resource-suffix
                  app.kubernetes.io/name: app-name
              maxSkew: 1
              topologyKey: topology.kubernetes.io/zone
              whenUnsatisfiable: ScheduleAnyway
            - labelSelector:
                matchLabels:
                  app.kubernetes.io/instance: RELEASE-NAME-resource-suffix
                  app.kubernetes.io/name: app-name
              maxSkew: 1
              topologyKey: kubernetes.io/hostname
              whenUnsatisfiable: ScheduleAnyway
          volumes:
            - configMap:
                name: RELEASE-NAME-resource-suffix-config-app
              name: config-app
            - name: config-app-secrets
              secret:
                defaultMode: 256
                secretName: RELEASE-NAME-resource-suffix-config-app-secrets
            - name: config-app-secrets-manual
              secret:
                secretName: RELEASE-NAME-resource-suffix-config-app-secrets-manual
            - name: config-hashed-aws-secret
              secret:
                secretName: RELEASE-NAME-resource-suffix-config-hashed-aws-secret-7a8f252
            - configMap:
                name: RELEASE-NAME-resource-suffix-config-hashed-cm-2b8747e
              name: config-hashed-cm
            - configMap:
                name: RELEASE-NAME-not-managed-cm
              name: config-not-managed-cm
            - name: config-not-managed-secret
              secret:
                secretName: RELEASE-NAME-not-managed-secret
            - configMap:
                name: RELEASE-NAME-resource-suffix-config-php
              name: config-php
            - configMap:
                defaultMode: 511
                name: RELEASE-NAME-resource-suffix-config-scripts
              name: config-scripts
            - emptyDir:
                sizeLimit: 10Mi
              name: ephemeral-app-cache
            - name: some-files
              persistentVolumeClaim:
                claimName: pvc-files
when multi deployment:
  1: |
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      labels:
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: app-name
        app.kubernetes.io/version: 0.1.0
      name: RELEASE-NAME-one
    spec:
      replicas: 1
      revisionHistoryLimit: 1
      selector:
        matchLabels:
          app.kubernetes.io/instance: RELEASE-NAME-one
          app.kubernetes.io/name: app-name
      template:
        metadata:
          labels:
            app.kubernetes.io/instance: RELEASE-NAME-one
            app.kubernetes.io/name: app-name
        spec:
          containers:
            - image: 123456.dkr.ecr.eu-central-1.amazonaws.com/test-server:0.1.0
              imagePullPolicy: IfNotPresent
              livenessProbe:
                httpGet:
                  path: /
                  port: app
              name: app
              ports:
                - containerPort: 80
                  name: app
                  protocol: TCP
              readinessProbe:
                httpGet:
                  path: /
                  port: app
              resources:
                requests:
                  cpu: 100m
                  memory: 128Mi
          enableServiceLinks: false
          nodeSelector:
            environment: test
          securityContext:
            sysctls:
              - name: net.ipv4.ip_unprivileged_port_start
                value: "0"
          serviceAccountName: RELEASE-NAME
          tolerations:
            - effect: NoSchedule
              key: environment
              operator: Equal
              value: test
          topologySpreadConstraints:
            - labelSelector:
                matchLabels:
                  app.kubernetes.io/instance: RELEASE-NAME-one
                  app.kubernetes.io/name: app-name
              maxSkew: 1
              topologyKey: topology.kubernetes.io/zone
              whenUnsatisfiable: ScheduleAnyway
            - labelSelector:
                matchLabels:
                  app.kubernetes.io/instance: RELEASE-NAME-one
                  app.kubernetes.io/name: app-name
              maxSkew: 1
              topologyKey: kubernetes.io/hostname
              whenUnsatisfiable: ScheduleAnyway
          volumes: null
  2: |
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      labels:
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: app-name
        app.kubernetes.io/version: 0.1.0
      name: RELEASE-NAME-two
    spec:
      replicas: 1
      revisionHistoryLimit: 1
      selector:
        matchLabels:
          app.kubernetes.io/instance: RELEASE-NAME-two
          app.kubernetes.io/name: app-name
      template:
        metadata:
          labels:
            app.kubernetes.io/instance: RELEASE-NAME-two
            app.kubernetes.io/name: app-name
        spec:
          containers:
            - image: 123456.dkr.ecr.eu-central-1.amazonaws.com/test-server:0.1.0
              imagePullPolicy: IfNotPresent
              livenessProbe:
                httpGet:
                  path: /
                  port: app
              name: app
              ports:
                - containerPort: 80
                  name: app
                  protocol: TCP
              readinessProbe:
                httpGet:
                  path: /
                  port: app
              resources:
                requests:
                  cpu: 100m
                  memory: 128Mi
          enableServiceLinks: false
          nodeSelector:
            environment: test
          securityContext:
            sysctls:
              - name: net.ipv4.ip_unprivileged_port_start
                value: "0"
          serviceAccountName: RELEASE-NAME
          tolerations:
            - effect: NoSchedule
              key: environment
              operator: Equal
              value: test
          topologySpreadConstraints:
            - labelSelector:
                matchLabels:
                  app.kubernetes.io/instance: RELEASE-NAME-two
                  app.kubernetes.io/name: app-name
              maxSkew: 1
              topologyKey: topology.kubernetes.io/zone
              whenUnsatisfiable: ScheduleAnyway
            - labelSelector:
                matchLabels:
                  app.kubernetes.io/instance: RELEASE-NAME-two
                  app.kubernetes.io/name: app-name
              maxSkew: 1
              topologyKey: kubernetes.io/hostname
              whenUnsatisfiable: ScheduleAnyway
          volumes: null
